### PR TITLE
[FEATURE: UIPT-21] - Recently viewed products

### PR DESCRIPTION
**Pull request by andriiveremchuk**

### CODE REVIEW

File: src/app/components/recently-viewed/recently-viewed.component.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    this.recentProductsSub = this.productService.storageSubject
+      .pipe(
+        takeWhile((data) => Boolean(data.length)),
+        concatMap(ids => ids),
+        map(productId => this.productService.getProductById(productId, ProductFormat.short)),
+        concatAll()
+      )
+      .subscribe(el => {
+      this.recentlyViewedProducts.push(el);
```
Comment (AlexLarusso) [inline src/app/components/recently-viewed/recently-viewed.component.ts]: please align

File: src/app/components/recently-viewed/recently-viewed.component.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    if (this.productService.recentlyViewed) {
+      this.productService.storageSubject.next(this.productService.recentlyViewed);
+    }
+
+    this.recentProductsSub = this.productService.storageSubject
+      .pipe(
+        takeWhile((data) => Boolean(data.length)),
+        concatMap(ids => ids),
+        map(productId => this.productService.getProductById(productId, ProductFormat.short)),
```
Comment (AlexLarusso) [inline src/app/components/recently-viewed/recently-viewed.component.ts]: We have agreed to use `getProductsByIds`.

File: src/app/components/recently-viewed/recently-viewed.component.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    const hours = 0.1; // 360 seconds
+    const now = new Date().getTime();
+    const hoursToMiliseconds = hours * 60 * 60 * 1000;
+    const TimerForRecentItemsExpiry = JSON.parse(localStorage.getItem('clearRecentlyViewed'));
+    const recentlyViewExists = JSON.parse(localStorage.getItem('recentlyViewed'));
+    const timeDifference = now - TimerForRecentItemsExpiry;
+
+    if (TimerForRecentItemsExpiry == null && recentlyViewExists) {
+      localStorage.setItem('clearRecentlyViewed', JSON.stringify(now));
+    } else {
```
Comment (AlexLarusso) [inline src/app/components/recently-viewed/recently-viewed.component.ts]: Will it work with `else if (condition2)` syntax?

File: src/app/components/recently-viewed/recently-viewed.html [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<div *ngIf='recentlyViewedProducts?.length' appScrollAnchor title="Recent Items">
+<div class="content__title">Recently viewed items</div>
+<app-product-carousel [productData]="recentlyViewedProducts"></app-product-carousel>
```
Comment (AlexLarusso) [inline src/app/components/recently-viewed/recently-viewed.html]: Please align

File: src/app/shared/product-item/product-item.component.ts [hunk 5]
```diff
@@ -17,4 +17,12 @@ export class ProductItemComponent {
 
   public faHeart = faHeart;
   public product: IProduct;
+
+  constructor(
+    private productService: ProductService,
+    ) { }
```
Comment (AlexLarusso) [inline src/app/shared/product-item/product-item.component.ts]: Align

File: src/app/shared/product-item/product-item.component.ts [hunk 6]
```diff
@@ -17,4 +17,12 @@ export class ProductItemComponent {
...
+
+  public OnClick(id: string) {
```
Comment (AlexLarusso) [inline src/app/shared/product-item/product-item.component.ts]: : void

File: src/app/shared/product-item/product-item.html [hunk 7]
```diff
@@ -1,6 +1,7 @@
 <div class="col-3 product-item__container">
   <div class="product-item__picture__container">
-    <a routerLink="/products/{{ productId }}">
+    <a routerLink="/products/{{ productId }}"
+       (click) = "OnClick(productId)">
```
Comment (AlexLarusso) [inline src/app/shared/product-item/product-item.html]: Remove spaces near "="

File: src/app/shared/services/product.service.ts [hunk 8]
```diff
@@ -48,6 +53,18 @@ export class ProductService {
...
   }
 
+  public addProductToLocalStorage(id: string): void {
+    this.recentlyViewed =
+      JSON.parse(localStorage.getItem('recentlyViewed')) || [];
+
+    if (this.recentlyViewed.indexOf(id) !== -1) {
+      this.recentlyViewed.splice(this.recentlyViewed.indexOf(id), 1);
+    }
+    this.recentlyViewed.unshift(id);
```
Comment (AlexLarusso) [inline src/app/shared/services/product.service.ts]: As I remember we have agreed to not use array ordering.

File: src/app/store/actions/web-storage.actions.ts [hunk 9]
```diff
@@ -6,14 +6,14 @@ export enum StorageActionTypes {
...
-  constructor(public payload: any){}
+  constructor(public payload: any) { }
 
   public static readonly TYPE = StorageActionTypes.SetValue;
   public type = SetValueToStorage.TYPE;
 }
 
 export class GetValueFromStorage implements Action {
-  constructor(public payload: any){}
+  constructor(public payload: any) { }
```
Comment (AlexLarusso) [inline src/app/store/actions/web-storage.actions.ts]: I think payload will be always string in both actions

File: src/app/store/reducers/web-storage.reducer.ts [hunk 10]
```diff
@@ -17,7 +17,9 @@ export function webStorageReducer(state = initialState, action: StorageAction):
...
         ...state,
         value: action.payload
-      }
+      };
     default: return state;
   }
-};
+}
+
+
```
Comment (AlexLarusso) [inline src/app/store/reducers/web-storage.reducer.ts]: Remove one empty

File: src/app/store/reducers/web-storage.reducer.ts [hunk 11]
```diff
@@ -17,7 +17,9 @@ export function webStorageReducer(state = initialState, action: StorageAction):
       return {
         ...state,
         value: action.payload
-      }
+      };
     default: return state;
```
Comment (AlexLarusso) [inline src/app/store/reducers/web-storage.reducer.ts]: Empty line before

File: src/app/components/recently-viewed/recently-viewed.component.ts (hunk 3 above)
Comment (andriiveremchuk) [reply src/app/components/recently-viewed/recently-viewed.component.ts]: Yup, its working with ```else if``` syntax

File: src/app/shared/services/product.service.ts (hunk 8 above)
Comment (andriiveremchuk) [reply src/app/shared/services/product.service.ts]: Correct, because array ordering is not reliable )

**Pull request merged**